part_test.go

```diff
@@ -427,6 +427,184 @@ func TestMultiSkipMalformedPart(t *testing.T) {
 	test.ContentContainsString(t, p.Content, want)
 }
 
+func TestReadPartErrorPolicy(t *testing.T) {
+	type testData struct {
+		policy                       enmime.ReadPartErrorPolicy
+		expectedPlainTextPartContent string
+		expectedHTMLPartContent      string
+		expectedErrorPlainTextPart   enmime.Error
+		expectedErrorHTMLPart        enmime.Error
+	}
+
+	var test0, test1, test2, test3 testData
+
+	// example policy 1: recover the buffer read
+	// only on base64.CorruptInputBuffer error and when content type is text/plain or text/html
+	examplePolicy1 := enmime.ReadPartErrorPolicy(func(p *enmime.Part, err error) bool {
+		allowedContentType := map[string]bool{
+			"text/html":  true,
+			"text/plain": true,
+		}
+		if enmime.IsBase64CorruptInputError(err) && allowedContentType[p.ContentType] {
+			return true
+		}
+		return false
+	})
+
+	// example policy 2: recover the buffer read
+	// only on base64.CorruptInputBuffer error and when content type is text/plain
+	examplePolicy2 := enmime.ReadPartErrorPolicy(func(p *enmime.Part, err error) bool {
+		allowedContentType := map[string]bool{
+			"text/plain": true,
+		}
+		if enmime.IsBase64CorruptInputError(err) && allowedContentType[p.ContentType] {
+			return true
+		}
+		return false
+	})
+
+	// example policy 3: always recover the buffer read, no matter the error
+	examplePolicy3 := enmime.ReadPartErrorPolicy(func(p *enmime.Part, err error) bool {
+		return true
+	})
+
+	plainContent := "Hello World!"
+	htmlContent := "<p>Hello World</p>"
+
+	ilegalBase64Error := enmime.Error{
+		Name:   enmime.ErrorMalformedBase64,
+		Detail: "illegal base64 data at input byte 0",
+		Severe: true,
+	}
+
+	keepingBufferWarning := enmime.Error{
+		Name:   enmime.ErrorMalformedChildPart,
+		Detail: "keeping the buffer read on error: illegal base64 data at input byte 0",
+		Severe: false,
+	}
+
+	test0 = testData{
+		policy:                       nil,
+		expectedPlainTextPartContent: "",
+		expectedErrorPlainTextPart:   ilegalBase64Error,
+		expectedHTMLPartContent:      "",
+		expectedErrorHTMLPart:        ilegalBase64Error,
+	}
+	test1 = testData{
+		policy:                       examplePolicy1,
+		expectedPlainTextPartContent: plainContent,
+		expectedErrorPlainTextPart:   keepingBufferWarning,
+		expectedHTMLPartContent:      htmlContent,
+		expectedErrorHTMLPart:        keepingBufferWarning,
+	}
+	test2 = testData{
+		policy:                       examplePolicy2,
+		expectedPlainTextPartContent: plainContent,
+		expectedErrorPlainTextPart:   keepingBufferWarning,
+		expectedHTMLPartContent:      "",
+		expectedErrorHTMLPart:        ilegalBase64Error,
+	}
+	test3 = testData{
+		policy:                       examplePolicy3,
+		expectedPlainTextPartContent: plainContent,
+		expectedErrorPlainTextPart:   keepingBufferWarning,
+		expectedHTMLPartContent:      htmlContent,
+		expectedErrorHTMLPart:        keepingBufferWarning,
+	}
+
+	tests := map[string]testData{
```

Comment:
You can combine map+struct literals to do this in one step that keeps the definition close to the name:
(from https://dave.cheney.net/2019/05/07/prefer-table-driven-tests)

```
func TestSplit(t *testing.T) {
    tests := map[string]struct {
        input string
        sep   string
        want  []string
    }{ 
        "simple":       {input: "a/b/c", sep: "/", want: []string{"a", "b", "c"}}, 
        "wrong sep":    {input: "a/b/c", sep: ",", want: []string{"a/b/c"}},
        "no sep":       {input: "abc", sep: "/", want: []string{"abc"}},
        "trailing sep": {input: "a/b/c/", sep: "/", want: []string{"a", "b", "c"}},
    }

    for name, tc := range tests {
        got := Split(tc.input, tc.sep)
        if !reflect.DeepEqual(tc.want, got) {
            t.Fatalf("%s: expected: %v, got: %v", name, tc.want, got)
        }
    }
}
```